Consolidating `ClientService` and js-ws and js-rpc integration

tegefaulkes opened the pull request:
### Description

In this PR we're consolidating the client service logic into a `ClientService` class that handles its life cycle.

### Issues Fixed

* Fixes #559 

### Tasks

- [x] 1. Create a ClientService class within the client domain.
- [x] 2. Move RPC and WebsocketServer creation and life-cycle into ClientService.
- [x] 3. PolykeyAgent will be updated to use this.
- [x] 4. Tests will be updated to use this. Should reduce setup logic.
- [x] 5. Add `PolykeyAgent.clientServiceHost`, `PolykeyAgent.clientServicePort`, `PolykeyAgent.agentServiceHost`, `PolykeyAgent.agentServicePort` as getter property so that they can be used directly instead of referring to encapsulated objects for the host and port.

### Final checklist

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* ~Squash and rebased~
* [x] Sanity check the final build

CMCDragonkai commented on src/vaults/VaultManager.ts:
```diff
@@ -36,6 +36,7 @@ import * as vaultsEvents from './events';
 import * as utils from '../utils';
 import * as vaultsUtils from '../vaults/utils';
 import * as vaultsErrors from '../vaults/errors';
+import * as vaultsEvents from './events';
```
This should be `./utils` and `./errors`.

tegefaulkes replied:
Fixed.

CMCDragonkai commented on tests/agent/handlers/notificationsSend.test.ts:
```diff
@@ -9,8 +9,8 @@ import os from 'os';
-import RPCClient from '@/rpc/RPCClient';
-import RPCServer from '@/rpc/RPCServer';
+import RPCClient from '@matrixai/rpc/dist/RPCClient';
+import RPCServer from '@matrixai/rpc/dist/RPCServer';
```
@addievo

```ts
import { RPCClient, RPCServer } from '@matrixai/rpc';
```

tegefaulkes replied:
Also fixed.

CMCDragonkai commented on tests/agent/handlers/nodesClaimsGet.test.ts:
```diff
@@ -7,10 +7,10 @@ import os from 'os';
 import * as nodesUtils from '@/nodes/utils';
 import { encodeProviderIdentityId } from '@/identities/utils';
-import RPCClient from '@/rpc/RPCClient';
-import RPCServer from '@/rpc/RPCServer';
```
Same here.

tegefaulkes replied:
Also fixed.

CMCDragonkai commented on tests/agent/handlers/nodesClosestLocalNode.test.ts:
```diff
@@ -7,8 +7,8 @@ import os from 'os';
-import RPCClient from '@/rpc/RPCClient';
-import RPCServer from '@/rpc/RPCServer';
+import RPCClient from '@matrixai/rpc/dist/RPCClient';
+import RPCServer from '@matrixai/rpc/dist/RPCServer';
```
All imports of `@matrixai/rpc`.

tegefaulkes replied:
Fixed up import of rpc.

amydevs committed:
fix: change client handlers to arrow function properties

amydevs committed:
feat: removing src/websockets and bringing in @matrixai/ws

amydevs committed:
wip: introducing `ClientService` and @matrixai/rpc

* `clientServiceHost`, `clientServicePort`, `agentServiceHost`, `agentServicePort` added to `PolykeyAgent``
* `ClientService` encapsulates both `WebSocketServer` and `RPCServer`

CMCDragonkai commented on src/nodes/utils.ts:
```diff
@@ -9,7 +9,7 @@ import * as nodesErrors from './errors';
 import { encodeNodeId, decodeNodeId } from '../ids';
 import { bytes2BigInt, never } from '../utils';
-import * as rpcErrors from '../rpc/errors';
+import * as rpcErrors from '@matrixai/rpc/dist/errors';
```
`import { errors as rpcErrors } from '@matrixai/rpc';`.

CMCDragonkai commented on src/nodes/agent/handlers/VaultsGitInfoGet.ts:
```diff
@@ -3,7 +3,7 @@ import type Logger from '@matrixai/logger';
 import type ACL from '../../../acl/ACL';
 import type VaultManager from '../../../vaults/VaultManager';
-import type { JSONRPCRequest } from '../../../rpc/types';
+import { JSONRPCRequest } from '@matrixai/rpc/dist/types';
```
`import type { JSONRPCRequest } from '@matrixai/rpc';`.

This is only possible if `@matrixai/rpc` does this:

```
export * from './types';
```

This is different from:

```
export * as types from './types';
```

CMCDragonkai commented on src/nodes/NodeConnectionManager.ts:
```diff
@@ -48,8 +48,8 @@ import * as validationUtils from '../validation/utils';
-import RPCServer from '../rpc/RPCServer';
-import * as rpcUtilsMiddleware from '../rpc/utils/middleware';
+import RPCServer from '@matrixai/rpc/dist/RPCServer';
+import * as rpcUtilsMiddleware from '@matrixai/rpc/dist/middleware';
```
`import { middleware as rpcMiddleware } from '@matrixai/rpc';`.

addievo committed:
wip : fixes all rpc imports, Fixes import orders
Lintfix

tegefaulkes committed:
tests: all nodes domain tests are working

Process is being held open for a time still, there is a timer not being cleaned up properly.

[ci skip]

tegefaulkes committed:
tests: fixed keys test

[ci skip]

tegefaulkes committed:
tests: fixed cert propagation tests

[ci skip]

tegefaulkes committed:
tests: moved and fixed agent handler tests

[ci skip]

tegefaulkes committed:
tests: fix for vaults cloning and pull problem

[ci skip]

tegefaulkes committed:
tests: removed unneeded test for `webcrypto`

[ci skip]

tegefaulkes committed:
tests: fixed IdentitiesManager test

[ci skip]

tegefaulkes committed:
tests: general fixes to tests.

[ci skip]

tegefaulkes committed:
wip: fixing up tests

[ci skip]

tegefaulkes committed:
tests: final test cleaning

[ci skip]

tegefaulkes committed:
Merge pull request #562 from MatrixAI/feature-test-fixes

Fixing `Polykey` and all tests for release

tegefaulkes committed:
tests: small fixes

[ci skip]

tegefaulkes committed:
lint: lintfix

[ci skip]

tegefaulkes committed:
test: fixing up tests

tegefaulkes committed:
test: small fix

[ci skip]

tegefaulkes committed:
dep: updating `@matrixai/rpc` to `^0.2.0`

[ci skip]

tegefaulkes committed:
fix: fixes for tests

All but 1 test fixed now.

[ci skip]

tegefaulkes committed:
dep: updated `@matrixai/ws` to `^1.1.5`

[ci skip]

tegefaulkes committed:
tests: fixed last test

[ci skip]

tegefaulkes committed:
fix: small fix to authenticationMiddleware.ts

tegefaulkes committed:
fix: fixed up `js-rpc` imports

CMCDragonkai commented on src/PolykeyAgent.ts:
```diff
@@ -40,6 +40,7 @@ import * as workersUtils from './workers/utils';
 import TaskManager from './tasks/TaskManager';
 import { serverManifest as clientServerManifest } from './client/handlers';
 import agentServerManifest from './nodes/agent/handlers';
+import ClientService from './client/ClientService';
```
Is this renaming of `import { serverManifest as clientServerManifest } from './client/handlers';` necessary?

I feel like the imports here can be tidied up too.

tegefaulkes replied:
It used to be, but now it's not needed.

CMCDragonkai replied:
Also I'm pretty sure that `client/handlers/index.ts` is supposed to default export the manifest.

Just like in `src/nodes/agent/handlers`.

The client manifest and server manifest are supposed to be default exports. Can you check?

tegefaulkes replied:
Yeah, that needs to be fixed as well. I made a note before somewhere. Ah, here https://github.com/MatrixAI/Polykey/pull/560#issuecomment-1752649971. 2. and 3. in that comment are no longer relevant.

There are something like 60 handlers, so it'll take too much time right now.

CMCDragonkai replied:
@addievo can do this, if you create a new issue, assign it to him.

tegefaulkes replied:
Import was fixed.

The handlers need to be addressed later. I'll make an issue in a moment.

tegefaulkes replied:
new issue at #563

tegefaulkes committed:
fix: fixed up `vaultManager` imports

CMCDragonkai commented on src/PolykeyClient.ts:
```diff
@@ -60,15 +62,17 @@ class PolykeyClient {
         clientUtilsMiddleware.middlewareClient(session),
         parserBufferByteLimit,
       ),
+      toError: networkUtils.toError,
```
I thought `toError` here would be specific to client domain? Or are we using `network/utils` for both client service and agent service?

tegefaulkes replied:
network domain is for stuff shared between client and agent network stuff.  One thing that still needs to be done is the `fromError` used for the clientService  needs to be a sensitive version that strips out the callstack.

CMCDragonkai replied:
The sensitive version is just whatever the `replacer` is. So I don't think `fromError` does any stripping. You define the `replacer` and that should be one stripping out the call stack and any other relevant information.

tegefaulkes replied:
`fromError` can do the stripping. Why do we even need the replacer? In any case, the `fromError` and `toError` needs to be fixed up. I didn't want to spend much more time on it with other things needing fixes at the time.

I think @amydevs should have a look over it.

CMCDragonkai replied:
It's cause `fromError` returns a JSON object to be stringified. The `replacer` can only be used when you do `JSON.stringify`. It's 2 steps.

tegefaulkes replied:
New issue at #564 to address this.

CMCDragonkai commented on src/PolykeyClient.ts:
```diff
@@ -1,18 +1,20 @@
 import { clientManifest } from './client/handlers/clientManifest';
+import * as networkUtils from './network/utils';
 
+let idGen: IdGen;
```
Why is `idGen` defined outside the entire class?

tegefaulkes replied:
I have no idea, that should not be done.

tegefaulkes replied:
It's not even defined anywhere. What is `IdGen` used for?

CMCDragonkai replied:
Yea, also I'm pretty the `idGen` is now optional. So for now just get rid of it.

CMCDragonkai replied:
By default removing `idGen` should be making it so that `id` field is always just `null`. As that is allowed in the JSON RPC spec.

CMCDragonkai replied:
In the future we can add an auto-increment. Or... even something useful for tracing.

tegefaulkes replied:
Removed.

tegefaulkes committed:
fix: review fixes

Fixed idGen usage in `PolykeyClient` and imports in `PolykeyAgent`
